src/main.cpp

```diff
@@ -830,13 +842,17 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
         view.SetBackend(viewMemPool);
 
         // do we already have it?
-        if (view.HaveCoins(hash))
+        if (view.HaveCoins(hash)) {
+            vHashTxnToUncache.push_back(hash);
```

Comment:
If I understand correctly, then we might remove this transaction from the cache even though there may be in-mempool spends of it, right?  I think we want some kind of check like you have in `TrimToSize`, where you only uncache if there's nothing in the mempool that spends it.

(Alternatively, if `HaveCoins` somehow reported whether any new caching happened, that might be helpful, but that seems like a harder change to make.)